### ISSUE TITLE
[Reimplementation] YOLOP - simaltaneous detection and semantic segmentation (not instance)

### ISSUE THREAD

**Issue by morkovka1337** [open-mmlab/mmdetection#9446]
### Prerequisite

- [X] I have searched [Issues](https://github.com/open-mmlab/mmdetection/issues) and [Discussions](https://github.com/open-mmlab/mmdetection/discussions) but cannot get the expected help.
- [X] I have read the [FAQ documentation](https://mmdetection.readthedocs.io/en/latest/faq.html) but cannot get the expected help.
- [X] The bug has not been fixed in the [latest version (master)](https://github.com/open-mmlab/mmdetection) or [latest version (3.x)](https://github.com/open-mmlab/mmdetection/tree/dev-3.x).

### 💬 Describe the reimplementation questions

Hello!

I'm trying to implement [YOLOP](https://github.com/hustvl/YOLOP) - a model that makes simaltaneous vehicle detection and drivable area and lane line segmentation. I know that from the architecture point of view it would be much better to split this model into two models - one for detection in the mmdetection and the other for segmentation in the mmsegmentation, but in such way the same picture will be passed through the backbone and neck twice. 

As far as I understood, I can not return segmentation mask as in the mmsegmentation in the shape of [B, num_classes, H, W]. Instead, for each detected class there should be a separate segmentation mask. Right now I've overcame this by adding empty masks to the detection results and then adding two bounding boxes of an image size with corresponding masks.
```
bbox_predict = self.bbox_head.predict(x_bbox, batch_data_samples, rescale=True)
lane_line_predict = self.lane_seg_head.predict(x_seg, batch_data_samples, rescale=True)
dr_area_predict = self.area_seg_head.predict(x_seg, batch_data_samples, rescale=True)
batch_data_samples = self.add_pred_to_datasample(batch_data_samples, bbox_predict)
batch_data_list = []
for batch_data, lane_line, dr_area in zip(batch_data_samples, lane_line_predict, dr_area_predict):
    seg_predictions = InstanceData(
        scores=torch.cat([batch_data.pred_instances.scores,
                torch.tensor([1, 1], device=batch_data.pred_instances.scores.device)]),
        labels=torch.cat([batch_data.pred_instances.labels,
                torch.tensor([1, 2], device=batch_data.pred_instances.labels.device)]),
        bboxes=torch.cat([batch_data.pred_instances.bboxes,
                torch.tensor([
                    [0, 0, *batch_data.ori_shape[::-1]],
                    [0, 0, *batch_data.ori_shape[::-1]]],
                device=batch_data.pred_instances.bboxes.device)]),
        masks = torch.zeros(len(batch_data.pred_instances.bboxes) + 2, *batch_data.ori_shape, device=batch_data.pred_instances.bboxes.device)
            )
    seg_predictions.masks[-1] = dr_area[1] > 0.5
    seg_predictions.masks[-2] = lane_line[1] > 0.5
    batch_data.pred_instances = seg_predictions
    batch_data_list.append(batch_data)
return batch_data_list
```

The question is: Is there any better way to return both object detecton and semantic segmentation predictions?

**Note that my work is based on the MMDet 3.x version.**

Appreciate any help, thanks!

### Environment

N\A

**Comment by Czm369** [comment]
Maybe you can refer to the implementation of HTC https://github.com/open-mmlab/mmdetection/blob/master/configs/htc/htc_r50_fpn_1x_coco.py

**Comment by morkovka1337** [comment]
> Maybe you can refer to the implementation of HTC https://github.com/open-mmlab/mmdetection/blob/master/configs/htc/htc_r50_fpn_1x_coco.py

Thanks, I will look there

**Comment by morkovka1337** [comment]
I've solved my problem. For those, who are interested:
* Detection head makes bounding box prediction (almost same as vanilla yolo); segmentation head predicts segmentation map, shape is [B, num_classes, H, W]
* In the model class iteself (which unites backbone, neck and heads) in the `predict` function I use PixelData (`from mmengine.structures import PixelData`) data structure. So the the whole predict is as follows:
```
def predict(self, batch_inputs: Tensor,
              batch_data_samples: SampleList) -> SampleList:
      """Predict results from a batch of inputs and data samples with post-
      processing."""
      x_seg, x_bbox = self.extract_feat(batch_inputs)
      bbox_predict = self.bbox_head.predict(x_bbox, batch_data_samples, rescale=True)
      lane_line_predict = self.lane_seg_head.predict(x_seg, batch_data_samples, rescale=True)
      dr_area_predict = self.area_seg_head.predict(x_seg, batch_data_samples, rescale=True)
      batch_data_samples = self.add_pred_to_datasample(
          batch_data_samples, bbox_predict)
      batch_data_list = []
      for batch_data, lane_line, dr_area in zip(batch_data_samples, lane_line_predict, dr_area_predict):
          batch_data.lane_line = PixelData(sem_seg=torch.max(lane_line, dim=0, keepdim=True).indices)
          batch_data.dr_area = PixelData(sem_seg=torch.max(dr_area, dim=0, keepdim=True).indices)
          batch_data_list.append(batch_data)
      return batch_data_list
```
* loss function is pretty easy, it returns a dictinoary of different loss component (bbox loss from bbox head + segmentation loss from segmentation head)
* For computing metrics from both bound boxes and masks I've created a new `metric` class which is inherited from CoCo and added mask IoU calculation there
* The only thing that concerns me is dataset class. It requires to read both detection bounding boxes and masks, so I attach masks a bit dirty to meta information. Dataset class is also inherited from CoCo.
